File: packages/pdf-gen-py/pdf_gen/utils/validate.py
from datetime import datetime, date
from typing import Any, Dict, List, Union

DateLike = Union[datetime, date, str]


def require_fields(data: Dict[str, Any], fields: List[str], template_name: str) -> None:
    missing = [f for f in fields if data.get(f) in (None, '')]
    if missing:
        raise ValueError(
            f'Missing required field(s) for "{template_name}" template: {", ".join(missing)}'
        )


def to_date(value: DateLike, fallback: datetime = None) -> datetime:
    if isinstance(value, datetime):
        return value
    if isinstance(value, date):
        return datetime(value.year, value.month, value.day)
    if isinstance(value, str):
        for fmt in ('%Y-%m-%d', '%Y-%m-%dT%H:%M:%S', '%Y-%m-%d %H:%M', '%m/%d/%Y'):
            try:
                return datetime.strptime(value, fmt)
            except ValueError:
                continue
    return fallback or datetime.now()


def format_date(value: DateLike) -> str:
    return to_date(value).strftime('%B %d, %Y')


def format_currency(amount: float, currency: str = 'USD') -> str:
    symbols = {'USD': '$', 'EUR': '€', 'GBP': '£', 'INR': '₹', 'PKR': 'Rs '}
    symbol = symbols.get(currency, currency + ' ')
    return f'{symbol}{amount:,.2f}'

File: packages/pdf-gen-js/src/utils/validate.ts
export function requireFields(data: Record<string, any>, fields: string[], templateName: string): void {
  const missing = fields.filter((f) => data[f] === undefined || data[f] === null || data[f] === '');
  if (missing.length > 0) {
    throw new Error(`Missing required field(s) for "${templateName}" template: ${missing.join(', ')}`);
  }
}

/** Accepts a Date instance or a date string/ISO string and returns a Date. */
export function toDate(value: Date | string | undefined, fallback: Date = new Date()): Date {
  if (value instanceof Date) return value;
  if (typeof value === 'string') {
    const parsed = new Date(value);
    if (!isNaN(parsed.getTime())) return parsed;
  }
  return fallback;
}

export function formatDate(value: Date | string | undefined): string {
  return toDate(value).toLocaleDateString(undefined, { year: 'numeric', month: 'long', day: 'numeric' });
}

export function formatCurrency(amount: number, currency: string = 'USD'): string {
  try {
    return new Intl.NumberFormat('en-US', { style: 'currency', currency }).format(amount);
  } catch {
    return `$${amount.toFixed(2)}`;
  }
}
